### ISSUE TITLE
Deploy on Docker Cloud using java alpine on Dockerfile causes problems on links/hostnames

### ISSUE THREAD

**Issue by marcelinobadin** [jhipster/generator-jhipster#3606]
##### **Overview of the issue**

I am trying to deploy my microservices stack using Docker Cloud. What I have notice is using Java Alpine in the Dockerfile causes problems to link applications (actually I am not using links, docker cloud permits to use hostnames directly). For example, linking my gateway application to http://jhipster-registry... makes my gateway to not be able to resolve this name.

Once I changed in the Dockerfile, from alpine to ```FROM java:8``` it worked fine.

This issue is more to share the experience.

Someone with more experience with docker can tell if it makes sense or was just luck on fixing my problems during the deploy.

**Comment by PierreBesson** [comment]
Hi @marcelinobadin,

Thanks for sharing your experience, don't worry this issue is relevant to us.

This might be related to the lack of support for DNS Search Path: [see alpine caveats docs](https://github.com/gliderlabs/docker-alpine/blob/master/docs/caveats.md).
It seems to be fixed in the development version. Can you try to use the following [Dockerfile](https://github.com/docker-library/openjdk/blob/8f8d04a5f77116be8ebfcaf84e4fcbd1190b95e8/8-jre/alpine/Dockerfile) to see if it fixes your issues (it's forked to use the alpine dev version).

**Comment by marcelinobadin** [comment]
I have tried, and it didn't work. I did this:

```
FROM alpine:3.3

ENV JHIPSTER_SLEEP 0
ENV LANG C.UTF-8

RUN { \
		echo '#!/bin/sh'; \
		echo 'set -e'; \
		echo; \
		echo 'dirname "$(dirname "$(readlink -f "$(which javac || which java)")")"'; \
	} > /usr/local/bin/docker-java-home \
	&& chmod +x /usr/local/bin/docker-java-home
ENV JAVA_HOME /usr/lib/jvm/java-1.8-openjdk/jre
ENV PATH $PATH:$JAVA_HOME/bin

ENV JAVA_VERSION 8u92
ENV JAVA_ALPINE_VERSION 8.92.14-r0

RUN set -x \
	&& apk add --no-cache \
		openjdk8-jre="$JAVA_ALPINE_VERSION" \
	&& [ "$JAVA_HOME" = "$(docker-java-home)" ]

# add directly the war
ADD *.war /app.war

RUN sh -c 'touch /app.war'
VOLUME /tmp
EXPOSE 8080
CMD echo "The application will start in ${JHIPSTER_SLEEP}s..." && \
    sleep ${JHIPSTER_SLEEP} && \
    java -Djava.security.egd=file:/dev/./urandom -jar /app.war
```

**Comment by PierreBesson** [comment]
@marcelinobadin Sorry, I gave you a wrong link. I wanted you to try to use `alpine:edge` instead of `alpine:3.3`.

**Comment by marcelinobadin** [comment]
It worked fine!

**Comment by PierreBesson** [comment]
Good news ! Since it doesn't affect to many users. I think we can just wait for alpine 3.4 to come out, then the problem will get fixed. People that really needs it can use "alpine:edge" or "java:8" if they want a stable version.

This problem is also affecting us for the Kubernetes sub-generator (coming soon, see #3443).

**Comment by marcelinobadin** [comment]
Well, apparently I missed something. Kubernets, as long as I could understand reading the documentation is pretty much what I'm looking for. If you need a alpha tester with your Kubernets implementation you can count on me.

Thanks for your help and time.

**Comment by marcelinobadin** [comment]
@PierreBesson another problem. The edge version of alpine worked for the same stack but cannot access hostnames on other stacks. On docker cloud, when you have services running in other stacks you have to put something like hostname.anotherstack to access. The alpine is not resolving this kind of hostname. I moved back to java:8 and it worked fine.

**Comment by PierreBesson** [comment]
So, service discovery for `serviceName.stackName` doesn't work for you with alpine images. This is probably normal as it is a docker-cloud specific feature. Maybe it relies on some things not available in alpine. If the problem persists for you, we will just have to write in the docs that this feature is not compatible with alpine for now...
But in this case use `java:8-jre`, it's half the size of `java:8` ([image layer comparison](https://imagelayers.io/?images=java:8,java:8-jre))


Did you try to deploy an alpine container outside of a stack and then link to it without the `.stackName`? I'm referring to this: https://docs.docker.com/docker-cloud/apps/service-links/#discovering-services-or-containers-not-included-in-a-stack Does this work ?

**Comment by marcelinobadin** [comment]
What I did was to use the full name (URL) of my service and it worked. It is an option since terminate a stack should be very rare.

Today I had experienced some instability with docker cloud service. I think I'm going to try Kubernets next week.

Many times my instances stop to "talk" with docker cloud, because of memory issues I think (i'm trying to use micro instances from AWS) and once it occurs everything stops and I have to terminate it and I loose data. I'm going to create a cluster with Galera. Did you heard about it?

Anyway, I will change to 9-jre and communication between slacks I will use the full hostname.

Thanks.
